### ISSUE TITLE
Pagination extremely slow for large data sets

### ISSUE THREAD

**Issue by applechief** [pallets/flask-sqlalchemy#518]
For a very large data set (millions of records in the table, 100k results) in a postgresql database, the count function is extremely slow.
The main query takes less than a second while the count is taking three minutes.

**Comment by ThiefMaster** [comment]
This is not really a Flask-SQLAlchemy problem but rather a common problem when doing pagination with COUNT and OFFSET: it works well for small/medium data sets, but is unsuitable for huge data sets not only due to COUNT being slow but also due to OFFSET being very inefficient when skipping lots of rows. BTHe advantage of using these two SQL constructs is, however, that it allows accessing any specific page immediately, which may be something you want if you paginate a small amount of data (take e.g. the pages in a forum thread as an example).

If you deal with huge data sets though, you usually want to allow only next/previous-page navigation since then you can use `somecol > x` filter and LIMIT which make efficient use of indexes. 

There is a very interesting post about this here: https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/
As you can see, there are different ways of doing it - having all of them available in Flask-SQLAlchemy would probably be overkill.. Keyset-style pagination as an alternative in Flask-SQLAlchemy could be nice though...

**Comment by fake-name** [comment]
There is a constrained case for this when you want to just be able to paginate the last **N** results along a column in a database. I recently had a project that exhibited this requirement. Basically, I want to limit the pagination max count to a hard-coded value. 

The underlying `Pagination()` class *supports this*, but for some reason it's not exposed through the `.pagination()` interface attached to `BaseQuery()`.

I wound up with a patched pagination call:

```
def limited_paginate(query_in, page=None, per_page=None, error_out=True, total_in=None):
	"""Returns ``per_page`` items from page ``page``.
	If no items are found and ``page`` is greater than 1, or if page is
	less than 1, it aborts with 404.
	This behavior can be disabled by passing ``error_out=False``.
	If ``page`` or ``per_page`` are ``None``, they will be retrieved from
	the request query.
	If the values are not ints and ``error_out`` is ``True``, it aborts
	with 404.
	If there is no request or they aren't in the query, they default to 1
	and 20 respectively.
	Returns a :class:`Pagination` object.
	"""

	if request:
		if page is None:
			try:
				page = int(request.args.get('page', 1))
			except (TypeError, ValueError):
				if error_out:
					abort(404)

				page = 1

		if per_page is None:
			try:
				per_page = int(request.args.get('per_page', 20))
			except (TypeError, ValueError):
				if error_out:
					abort(404)

				per_page = 20
	else:
		if page is None:
			page = 1

		if per_page is None:
			per_page = 20

	if error_out and page < 1:
		abort(404)

	items = query_in.limit(per_page).offset((page - 1) * per_page).all()

	if not items and page != 1 and error_out:
		abort(404)

	# No need to count if we're on the first page and there are fewer
	# items than we expected.
	if page == 1 and len(items) < per_page:
		total = len(items)
	elif total_in:
		total = total_in
	else:
		total = query_in.order_by(None).count()

	return Pagination(query_in, page, per_page, total, items)
```

It seems like `BaseQuery(),paginate()` should really allow you to specify a `total` parameter.

**Comment by iandees** [comment]
It looks like #613 was merged to support this. Could we get a new release so we can pull this in to our software please?

**Comment by rsyring** [comment]
Looks like #613 could get ported to a 2.x release.

**Comment by zhuxiaochu** [comment]
It is a good idea.Then we can even make the most use of counter table by **total_in**.Large data sets will not be a problem.

**Comment by mvanbaak** [comment]
Pagination based on something like https://github.com/djrobstep/sqlakeyset would be very nice to have.

**Comment by davidism** [comment]
I think this has been addressed in 2.4. However, it's just hard to do this in a general way that supports all use cases. If you have very large data sets and are seeing performance issues, it might be time to use custom pagination. There are also various other issues and PRs open about refactoring pagination that might address this.
